crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -362,17 +362,20 @@ impl<N: ProviderNodeTypes> HeaderProvider for BlockchainProvider2<N> {
     }
 
     fn header_td_by_number(&self, number: BlockNumber) -> ProviderResult<Option<U256>> {
-        // If the TD is recorded on disk, we can just return that
-        if let Some(td) = self.database.header_td_by_number(number)? {
-            Ok(Some(td))
-        } else if self.canonical_in_memory_state.hash_by_number(number).is_some() {
-            // Otherwise, if the block exists in memory, we should return a TD for it.
+        if self.canonical_in_memory_state.hash_by_number(number).is_some() {
+            // If the block exists in memory, we should return a TD for it.
             //
             // The canonical in memory state should only store post-merge blocks. Post-merge blocks
             // have zero difficulty. This means we can use the total difficulty for the last
-            // persisted block number.
-            let last_persisted_block_number = self.database.last_block_number()?;
-            self.database.header_td_by_number(last_persisted_block_number)
+            // finalized block number, so that we are not affected by reorgs.
+            let last_finalized_num_hash = self
+                .canonical_in_memory_state
+                .get_finalized_num_hash()
+                .unwrap_or_else(|| BlockNumHash::new(0, B256::default()));
+            self.database.header_td_by_number(last_finalized_num_hash.number)
+        } else if let Some(td) = self.database.header_td_by_number(number)? {
```

Comment:
we won't ever hit that if we hardcoded the post merge ttd

but we can't use `BlockNumHash::new(0, B256::default())` because here because this would return the genesis ttd.
this would only be reachable in custom setups, but the unwrap_or_else should use the `last_block_number()` instead, even if there's a race condition, but this only exists until there is a finalized block, so imo this is fine

Reply:
> i think, if its not in memory we can just return whatever self.database method returns

good point, done here https://github.com/paradigmxyz/reth/pull/11382/commits/36b11f7683f416519bbd871d7c798239d2059dce

Reply:
> but we can't use `BlockNumHash::new(0, B256::default())` because here because this would return the genesis ttd.
> this would only be reachable in custom setups, but the unwrap_or_else should use the `last_block_number()` instead, even if there's a race condition, but this only exists until there is a finalized block, so imo this is fine

yep, if `get_finalized_num_hash` returns `None` and we try to get `last_block_number` the test is flaky again, should we do that instead?

Reply:
I think the flaky test is fine, we know why it's flaky and why this isn't an issue.
also total difficulty will be removed from rpc so this case will eventually go away